scripts/release/tag-images.py

```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env python3
+"""Tags images used in a particular release and pushes the tags into
+our local docker registry at docker.montagu.dide.ic.ac.uk:5000.  Use
+tag 'latest' to select the most recent conforming git tag (this will
+not set things to be the docker 'latest' tag though).  If run with the
+"publish" option it will also publish images to
+https://hub.docker.com/u/vimc
+
+Usage:
+  tag-images.py tag [--publish] <version>
+  tag-images.py publish <version>
+
+"""
+import docker
+import docopt
+from subprocess import run
+import os
+import re
+import git_helpers
+from helpers import get_latest_release_tag, validate_release_tag
+
+# This feels like something we should have elsewhere; it's a map of
+# the name of the *repo* (the key here) with the name of the submodule
+# *subdirectory* (the value, which then maps onto the docker compose
+# container name).  It's easy enough to lift this out later though.
+#
+# Not currently included are montagu-portal-integration-tests and
+# montagu-api-blackbox-tests (both of which version against the api
+# submodule).
+container_repo_map = {
+    "montagu-admin-portal": "admin-portal",
+    "montagu-api": "api",
+    "montagu-cert-tool": "cert-tool",
+    "montagu-cli": "api",
+    "montagu-contrib-portal": "contrib-portal",
+    "montagu-db": "db",
+    "montagu-migrate": "db",
+    "montagu-orderly": "orderly",
+    "montagu-report-portal": "report-portal",
+    "montagu-reporting-api": "reporting-api",
+    "montagu-reverse-proxy": "proxy"
+}
+
+registry_local = "docker.montagu.dide.ic.ac.uk:5000"
+registry_hub = "vimc"
+
+def set_image_tag(name, version):
+    repo_name = container_repo_map[name]
+    sha = git_helpers.get_past_submodule_version(repo_name, version)
+    d = docker.client.from_env()
+    img = d.images.pull("{}/{}:{}".format(registry_local, name, sha))
+    tag_and_push(img, registry_local, name, version)
+
+def set_image_tags(version):
+    print("Setting image tags")
+    for name in container_repo_map.keys():
+        print("  - " + name)
+        set_image_tag(name, version)
+
+def publish_images(version):
+    d = docker.client.from_env()
+    print("Pushing release to docker hub")
+    for name in container_repo_map.keys():
+        img = d.images.get("{}/{}:{}".format(registry_local, name, version))
+        publish_image(img, name)
+
+def publish_image(img, name):
+    tags = img.tags
+    published = [x.split(":")[-1] for x in tags if x.startswith(registry_hub)]
+    existing = [x.split(":")[-1] for x in tags if x.startswith(registry_local)]
```

Comment:
I feel like this could be more readable. Maybe replace `x.split(":")[-1]` with a named function (`get_version_from_docker_tag`) or even a class. e.g.

```
class DockerTag:
    def __init__(self, raw):
        parts = raw.split("/")
        self.registry = parts[0]
        parts = parts[1].split(":")
        self.name = parts[0]
        self.version = parts[1]

...

published = [t.version for t in [DockerTag(x) for x in tags] if t.registry == registry_hub]
```

Reply:
I've gone with the class approach and also eliminated some repetitive formatting operations